File: src/normal_solver.py
import sys
import json
import solver


def Solve(task):
    if not solver.Solvable(json.dumps(task["board"])):
        return

    for maxSteps in range(task["maxSteps"]+1):
        curTask = {"board": task["board"],
                   "steps": task["steps"],
                   "maxSteps": maxSteps}
        result = json.loads(solver.Solve(json.dumps(curTask)))
        if result != None:
            return result


if __name__ == "__main__":
    task = json.loads(sys.stdin.read())
    result = Solve(task)
    print(json.dumps(result))

File: src/normal_solver.cpp
#include "solver.hpp"
#include "adapter.hpp"
#include <iostream>

std::optional<std::vector<puzzle::Direction>> Solve(const Task& task)
{
    for (int maxSteps = 0;maxSteps <= task.maxSteps;++maxSteps)
    {
        auto historySteps = task.steps;
        auto steps = puzzle::Solver<5, 5>::Solve(task.board, historySteps, maxSteps);
        if (steps)
            return steps;
    }
    return std::nullopt;
}

int main()
{
    auto task = json::parse(ReadAll(std::cin)).get<Task>();
    bool solvable = puzzle::Solver<5, 5>::Solvable(task.board);
    auto steps = solvable ? Solve(task) : std::nullopt;
    std::cout << json(steps);
    return 0;
}